### ISSUE TITLE
make: use of immediate value of variables before they have their final value

### ISSUE THREAD

**Issue by cladmi** [RIOT-OS/RIOT#8913]
#### Description

Some makefiles use variables with their immediate value before they get the final one.
It's not problematic for build flags but for variables containing files/path.

The root of the problem is that the makefiles do not follow a pattern of defining all variables first/including sub makefiles, and then only after define the rules using the variables as file target or file dependency.

This issue will be the central one to reference sub pull requests to fix this uses.

#### Debugging

The problem can be seen by using `$(info debug VARIABLE $(VARIABLE))` before usage.

##### Makefile.include

In `Makefile.include`, all uses of `$(BASELIBS)`, except when defining `_LINK`, are using the immediate value of `$(BASELIBS)` which does not have its final value as this is done by including `modules.inc.mk` 200 lines below.

All the uses are currently solved by hidden "tricks"

* `$(ELFFILE): $(BASELIBS)` which works because there is a `$(ELFFILE): FORCE` rule, so it will be rebuilt anyway, and `_LINK` is using the deferred value of `BASELIBS`.
* `link: $(BASELIBS)` when in the `RIOTNOLINK` case, where BASELIBS is missing `$(USEPKG:%=$(BINDIR)/%.a)` files, but as `$(BINDIR)/$(APPLICATION_MODULE).a` has a dependency to `$(USEPKG:%=$(BINDIR)/%.a)` they are still build anyway.
* The `clean` before building trick for parallel execution is using `$(BASELIBS)` also without the `$(USEPKG:%=$(BINDIR)/%.a)` files. But as there is the `RIOTPKG/%/Makefile.include` target also depending on `clean`, the packages are built after `clean`.

##### pkg/ `PKG_BUILDDIR`

Some packages makefile use `$(PKG_BUILDDIR)` with its immediate value when defining targets before it is defined by including `pkg/pkg.mk`.

It is mitigated because then they are treated as non file targets, and because the command repeats the variable and does not use the automatic make variables like `$@`. https://github.com/RIOT-OS/RIOT/pull/8509#discussion_r178037265

**Comment by cladmi** [comment]
Includes is also evaluated directly https://github.com/RIOT-OS/RIOT/blob/1f7ec9b2082eec7419f50a7f35eefc84b26a6e90/makefiles/libc/newlib.mk#L89 even if `CFLAGS_FPU` can be changed later by `pkg/nordic_softdevice_ble` which would change the `INCLUDES` when using `llvm` https://github.com/RIOT-OS/RIOT/blob/1f7ec9b2082eec7419f50a7f35eefc84b26a6e90/makefiles/toolchain/llvm.inc.mk#L70

In that case it is not breaking as llvm is said incompatible but it shows an issue with these immediate evaluations.

**Comment by cladmi** [comment]
This one is also related, where a file that should be put in `BUILDDEPS` is declared by a makefile included after using `BUILDDEPS`: https://github.com/RIOT-OS/RIOT/pull/10492#issuecomment-444108243

It goes with many definitions orders in our Makefiles that should be addressed in a deterministic way at some point.
